### PR TITLE
Wayland related modifications

### PR DESCRIPTION
**Pull request by xfangfang**
The content of this PR has only been tested under textInputV3, as my understanding of Wayland is limited, so it may contain issues.

## 1. Improves

The current Wayland implementation defaults to automatically enabling the input method when the application starts, leading to:

- Issue 1: Unable to normally use application shortcuts: https://github.com/xfangfang/wiliwili/issues/306
- Issue 2: When using the on-screen keyboard, the keyboard is forcibly enabled and cannot be turned off

Therefore, I directly removed the keyboard activation functions of `textInputV1Enter` and `textInputV3Enter`, and at the same time, simply implemented the IME status-related functions, allowing users to enable or disable the keyboard through `glfwSetInputMode`.

## 2. Fix for endless preedit cursor update

While implementing the above, I found that `textInputV3Done` is called repeatedly after the application starts. 

After reading the documentation:

> * When the client receives a done event with a serial different
> * than the number of past commit requests, it must proceed with
> * evaluating and applying the changes as normal, except it should
> * not change the current state of the zwp_text_input_v3 object.

I made a simple adjustment, calling `zwp_text_input_v3_set_cursor_rectangle` / `zwp_text_input_v1_set_cursor_rectangle` only when the coordinates change. As I tested with an on-screen keyboard, I am not sure if it will affect the other input method, but it did indeed stop the loop of `textInputV3Done` being called repeatedly.

### CODE REVIEW

**Comment by ashie** [comment]
> While implementing the above, I found that `textInputV3Done` is called repeatedly after the application starts.
> 
> After reading the documentation:
> 
> > * When the client receives a done event with a serial different
> > * than the number of past commit requests, it must proceed with
> > * evaluating and applying the changes as normal, except it should
> > * not change the current state of the zwp_text_input_v3 object.

ref: https://wayland.app/protocols/text-input-unstable-v3#zwp_text_input_v3:event:done

**Comment by ashie** [comment]
Thanks for your feedback!

> 2. Fix for endless preedit cursor update

The fix for this issue looks good.

> 1. Improves

I have some concerns on this fix.
It seems that it introduces different UX with other platform especially Windows.
(We consider that the Windows backend is the reference implementation of multi-platform IM feature.)
In this context, IM can be always enabled or disabled by one of these elements:

* Short cut keys
* Panel UI provided by IM frame work
* Application

and the status should be synchronized no matter what element you use.
But this fix doesn't seem synchronized, and only the application can enable IM at start up.

**Comment by xfangfang** [comment]
> It seems that it introduces different UX with other platform especially Windows.

This is indeed a problem. I have only tested the touchscreen experience in the Wayland desktop environment (SteamDeck) of KDE, and have not found a way to manually turn off the on-screen keyboard. I have also not tested the support for other input methods.

I may install Wayland on my new computer for more detailed testing in about a week or two.

---
Updated: 02/19/2025

Sorry I'm not very familiar with Wayland and didn't found a function to monitor the input method status.

**Comment by ashie** [comment]
Sorry for not responding for a while.

Although current `glfwSetInputMode(win, GLFW_IME, value)` implementation intends to change platform's IME status as I described before, I also know that there is a requirement that want bypassing IME like your patch. To resolve both requirements, it would be better to extend the 3rd argument like the following:

* `GLFW_IME_DISABLE`: Force bypass IME. Should be supported on all platforms.
* `GLFW_IME_ENABLE`: Don't bypass IME. Should be supported on all platforms
* `GLFW_IME_OFF`: Turn off IME by platform's fashion. Might not be supported on some platforms. 
* `GLFW_IME_ON`: Turn on IME by platform's fashion. Might not be supported on some platforms.

We need to consider the detailed specifications a bit more though.

**Comment by xfangfang** [comment]
Thank you very much for your reply. I look forward to the future updates. You can close this PR at any time.

**Comment by ghostflyby** [comment]
> Sorry for not responding for a while.
> 
> Although current `glfwSetInputMode(win, GLFW_IME, value)` implementation intends to change platform's IME status as I described before, I also know that there is a requirement that want bypassing IME like your patch. To resolve both requirements, it would be better to extend the 3rd argument like the following:
> 
> * `GLFW_IME_DISABLE`: Force bypass IME. Should be supported on all platforms.
> * `GLFW_IME_ENABLE`: Don't bypass IME. Should be supported on all platforms
> * `GLFW_IME_OFF`: Turn off IME by platform's fashion. Might not be supported on some platforms.
> * `GLFW_IME_ON`: Turn on IME by platform's fashion. Might not be supported on some platforms.
> 
> We need to consider the detailed specifications a bit more though.

I‘m confused by the naming. We have two toggles `IME/composition with candidate window` and `ASCII text input`, what's their Truth table on the four GLFW status?

In the other word, what do you mean by `force bypass`, is there still ASCII text input?

**Comment by ashie** [comment]
FYI, I've pushed a new prototype branch:

https://github.com/clear-code/glfw/tree/im-support-text-input-focus

For applications affected by the issue discussed here, the expected integration would be to explicitly manage text input focus. In practice, that means calling:

```c
glfwSetTextInputFocus(window, GLFW_FALSE);
```

when the application starts or enters a non-text-input mode, and switching to `GLFW_TRUE` only while a text input widget (chat box, search field, editor, etc.) is active.

While revisiting the IME work, I ended up changing the design direction somewhat. Instead of treating this primarily as IME enable/disable, the branch introduces explicit text input focus management via `glfwSetTextInputFocus()`.

The idea is that applications explicitly tell GLFW when they enter or leave a text input context, rather than GLFW trying to infer this from native window focus alone.

For applications that use `glfwSetTextInputFocus()`, the original goal of this PR is largely covered:

* Wayland text-input is no longer active all the time.
* Applications can keep text-input disabled during gameplay or other non-text-input interaction.
* Text-input can be enabled only while text input is actually expected.
* This avoids IME/text-input interference with shortcuts and similar key handling.

One important difference is that existing applications keep the previous behavior by default for compatibility. The new behavior becomes active only after an application opts into explicit text input focus management.

In other words, the problem is now modeled as "does the application currently have a text input context?" rather than "is IME enabled or disabled?".

The IME status control parts of this PR are intentionally not covered at the moment, as I currently consider them a separate concern.

I'm still testing and refining things, but I wanted to share the current direction since this discussion was one of the things that led me to rethink the design.

**Comment by ashie** [comment]
> FYI, I've pushed a new prototype branch:
> 
> https://github.com/clear-code/glfw/tree/im-support-text-input-focus

I've merged into https://github.com/glfw/glfw/pull/2130.
Thanks for your contribution!
